File: unity/CommandWarfare/Assets/Scripts/Core/Combat/DamageTypes.cs
using System;
using System.Collections.Generic;
using CommandWarfare.Data;

namespace CommandWarfare.Core.Combat
{
    /// <summary>Port of play/shared/damageTypes.ts — type matchup bonus damage.</summary>
    public static class DamageTypes
    {
        public readonly struct BonusRule
        {
            public string[] AttackerTags { get; }
            public string[] DefenderTags { get; }
            public int BonusDamage { get; }

            public BonusRule(string[] attackerTags, string[] defenderTags, int bonusDamage)
            {
                AttackerTags = attackerTags;
                DefenderTags = defenderTags;
                BonusDamage = bonusDamage;
            }
        }

        /// <summary>Intentionally empty until card tags are audited (matches TS stub).</summary>
        public static readonly BonusRule[] Rules = Array.Empty<BonusRule>();

        public static List<string> TagsFrom(CardDefinition card)
        {
            var tags = new List<string>();
            if (card == null) return tags;
            if (!string.IsNullOrEmpty(card.primaryType)) tags.Add(card.primaryType);
            if (!string.IsNullOrEmpty(card.secondaryType)) tags.Add(card.secondaryType);
            if (card.keywords != null)
                tags.AddRange(card.keywords);
            return tags;
        }

        public static List<string> TagsFrom(UnitTokenSnapshot snap)
        {
            var tags = new List<string>();
            if (snap.PrimaryType != null) tags.Add(snap.PrimaryType);
            if (snap.SecondaryType != null) tags.Add(snap.SecondaryType);
            if (snap.Keywords != null)
                tags.AddRange(snap.Keywords);
            return tags;
        }

        public static int Bonus(IReadOnlyList<string> attackerTags, IReadOnlyList<string> defenderTags)
        {
            if (Rules.Length == 0 || attackerTags == null || defenderTags == null) return 0;

            var atk = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            var def = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            foreach (var t in attackerTags) atk.Add(t);
            foreach (var t in defenderTags) def.Add(t);

            var bonus = 0;
            foreach (var rule in Rules)
            {
                var atkHit = false;
                foreach (var t in rule.AttackerTags)
                {
                    if (atk.Contains(t)) { atkHit = true; break; }
                }
                if (!atkHit) continue;

                var defHit = false;
                foreach (var t in rule.DefenderTags)
                {
                    if (def.Contains(t)) { defHit = true; break; }
                }
                if (defHit) bonus += rule.BonusDamage;
            }
            return bonus;
        }
    }

    /// <summary>Minimal card fields for damage type tagging without full catalog lookup.</summary>
    public readonly struct UnitTokenSnapshot
    {
        public string PrimaryType { get; }
        public string SecondaryType { get; }
        public IReadOnlyList<string> Keywords { get; }

        public UnitTokenSnapshot(string primaryType, string secondaryType, IReadOnlyList<string> keywords)
        {
            PrimaryType = primaryType;
            SecondaryType = secondaryType;
            Keywords = keywords;
        }
    }
}

File: play/shared/damageTypes.ts
/**
 * Damage-vs-type bonuses (stub).
 *
 * Future: spear/pike vs Mounted, anti-armor vs Heavy, etc.
 * Combat should call damageTypeBonus() from strikeDamage once the matrix is filled.
 */

export type DamageTypeTag =
  | 'Mounted'
  | 'Siege'
  | 'Infantry'
  | 'Spear'
  | 'Pike'
  | 'Heavy'
  | string

export type DamageTypeBonusRule = {
  /** Attacker must have at least one of these tags. */
  attackerTags: DamageTypeTag[]
  /** Defender must have at least one of these tags. */
  defenderTags: DamageTypeTag[]
  bonusDamage: number
}

/**
 * Example rules (not active until populated into DAMAGE_TYPE_RULES):
 * - { attackerTags: ['Spear', 'Pike'], defenderTags: ['Mounted'], bonusDamage: 1 }
 */
export const DAMAGE_TYPE_RULES: DamageTypeBonusRule[] = [
  // Intentionally empty until card tags are audited.
]

/** Collect combat type tags from a card/unit snapshot. */
export function damageTypeTagsFrom(card: {
  primaryType?: string | null
  secondaryType?: string | null
  keywords?: string[] | null
}): DamageTypeTag[] {
  const tags: DamageTypeTag[] = []
  if (card.primaryType) tags.push(card.primaryType)
  if (card.secondaryType) tags.push(card.secondaryType)
  for (const k of card.keywords ?? []) tags.push(k)
  return tags
}

/**
 * Extra damage from type matchups. Returns 0 until DAMAGE_TYPE_RULES is populated.
 */
export function damageTypeBonus(
  attackerTags: DamageTypeTag[],
  defenderTags: DamageTypeTag[],
): number {
  if (!DAMAGE_TYPE_RULES.length) return 0
  const atk = new Set(attackerTags.map((t) => t.toLowerCase()))
  const def = new Set(defenderTags.map((t) => t.toLowerCase()))
  let bonus = 0
  for (const rule of DAMAGE_TYPE_RULES) {
    const atkHit = rule.attackerTags.some((t) => atk.has(t.toLowerCase()))
    const defHit = rule.defenderTags.some((t) => def.has(t.toLowerCase()))
    if (atkHit && defHit) bonus += rule.bonusDamage
  }
  return bonus
}
